client/python/gradio_client/utils.py

```diff
@@ -297,13 +299,119 @@ async def get_pred_from_ws(
+                stream_sse(
+                    client, data, hash_data, helper, sse_url, sse_data_url, cookies
+                )
+            ),
+        ],
+        return_when=asyncio.FIRST_COMPLETED,
+    )
+
+    for task in pending:
+        task.cancel()
+        try:
+            await task
+        except asyncio.CancelledError:
+            pass
+
+    assert len(done) == 1
+    for task in done:
+        return task.result()
+
+
+async def check_for_cancel(helper: Communicator, cookies: dict[str, str] | None):
+    while True:
+        await asyncio.sleep(0.05)
+        with helper.lock:
+            if helper.should_cancel:
+                break
+    if helper.event_id:
+        async with httpx.AsyncClient() as http:
+            await http.post(
+                helper.reset_url, json={"event_id": helper.event_id}, cookies=cookies
+            )
+    raise CancelledError()
+
+
+async def stream_sse(
+    client: httpx.AsyncClient,
+    data: dict,
+    hash_data: dict,
+    helper: Communicator,
+    sse_url: str,
+    sse_data_url: str,
+    cookies: dict[str, str] | None = None,
+) -> dict[str, Any]:
+    try:
+        async with client.stream(
+            "GET", sse_url, params=hash_data, cookies=cookies
+        ) as response:
+            async for line in response.aiter_text():
+                if line.startswith("data:"):
+                    resp = json.loads(line[5:])
+                    with helper.lock:
+                        has_progress = "progress_data" in resp
+                        status_update = StatusUpdate(
+                            code=Status.msg_to_status(resp["msg"]),
+                            queue_size=resp.get("queue_size"),
+                            rank=resp.get("rank", None),
+                            success=resp.get("success"),
+                            time=datetime.now(),
+                            eta=resp.get("rank_eta"),
+                            progress_data=ProgressUnit.from_msg(resp["progress_data"])
+                            if has_progress
+                            else None,
+                        )
+                        output = resp.get("output", {}).get("data", [])
+                        if output and status_update.code != Status.FINISHED:
+                            try:
+                                result = helper.prediction_processor(*output)
+                            except Exception as e:
+                                result = [e]
+                            helper.job.outputs.append(result)
+                        helper.job.latest_status = status_update
+
+                    if resp["msg"] == "queue_full":
+                        raise QueueError("Queue is full! Please try again.")
+                    elif resp["msg"] == "send_data":
+                        event_id = resp["event_id"]
+                        helper.event_id = event_id
+                        async with httpx.AsyncClient() as http:
```

Comment:
Nice. Can we use the same client as the one that's streaming or do we need to create a new one?